### DISCUSSION TITLE
How to halt script execution?

### DISCUSSION THREAD

**Discussion by gilmatok** [sebastienros/jint#1058] [Q&A]
I'm using this to make a scripting interface for a MMORPG game server.

Once the user initates a chat with a NPC, the script is executed. For example, I can open up a Yes/No window for the NPC using this:

var yes = cm.SendYesNo("yes or no?");

if (yes)
{
// do smthn
} else {
// do smthn else
}

However, I want to halt the exceution of the script after I send SendYesNo and continue it once I return the boolean that is returned. How can I do that code-wise?

**Comment by doronguttman** [comment]
either make cm.SendYesNo thread blocking or use a callback

**Comment by gilmatok** [comment]
```
private TaskCompletionSource<bool> response;
private string _text = "";

public void SetResult(bool result)
{
    this.response.SetResult(result);
}

public async Task<bool> SendYesNo()
{
    response = new TaskCompletionSource<bool>();

    MessageBox.Show(this._text, "", MessageBoxButtons.YesNo);

    this._text = "";

    return response.Task.Result;
}
```

I got around it using this code, however I can't call the await keyword using JavaScript. Any suggestions?

**Comment by doronguttman** [comment]
I think you over engineered this.
Just do this:
public bool SendYesNo(string message)
{
    return MessageBox.Show(message, "", MessageBoxButtons.YesNo) ==
DialogResult.Yes;
}

**Comment by gilmatok** [comment]
The MessageBox was just an example, I'm not actually using MessageBoxes. This is for a MMORPG, so the MessageBoxes are imaginary NPC chats.

**Comment by doronguttman** [comment]
ok, in that case, use the taskcompletionsource, but I didn't see anywhere
in your code snippet where you are calling SetResult.
In any case, I don't think Jint supports await, so just do:

public bool SendYesNo(string message)
{
    var tcs = new TaskCompletionSource<bool>();
    Task.Factory.StartNewTask(() =>
    {
         // do what needs to be done in order to get the result back

```
     // when finished call this:
     tcs.TrySetResult(result);
}

return tcs.Task.Result;
```

}

**Comment by gilmatok** [comment]
```
    private TaskCompletionSource<bool> tcs;

    public bool SendYesNo()
    {
        tcs = new TaskCompletionSource<bool>();
        Task.Factory.StartNew(() =>
        {
            MessageBox.Show("aeae", "", MessageBoxButtons.YesNo);
        });

        return tcs.Task.Result;
    }

    public void SetResult(bool result)
    {
        tcs.TrySetResult(result);
    }
```

Am I doing something wrong? By the way, I have to have "SetResult" in a separate method, as I'm calling it from another class.

**Comment by doronguttman** [comment]
if that is the case I would do this:

```
private TaskCompletionSource<bool> _tcs = new TaskCompletionSource<bool>();

public bool SendYesNo(string message)
{
    Task.Factory.StartNew(() =>
    {
        SomeOtherClass.SomeStaticMethod(message, this);
    });

    return _tcs.Task.Result;
}

public void SetResult(bool result)
{
    _tcs.TrySetResult(result);
}
```
